urllib3/contrib/_securetransport/low_level.py

```diff
@@ -0,0 +1,343 @@
+"""
+Low-level helpers for the SecureTransport bindings.
+
+These are Python functions that are not directly related to the high-level APIs
+but are necessary to get them to work. They include a whole bunch of low-level
+CoreFoundation messing about and memory management. The concerns in this module
+are almost entirely about trying to avoid memory leaks and providing
+appropriate and useful assistance to the higher-level code.
+"""
+import base64
+import ctypes
+import itertools
+import re
+import os
+import ssl
+import tempfile
+
+from .bindings import Security, CoreFoundation, CFConst
+
+
+# This regular expression is used to grab PEM data out of a PEM bundle.
+_PEM_CERTS_RE = re.compile(
+    b"-----BEGIN CERTIFICATE-----\n(.*?)\n-----END CERTIFICATE-----", re.DOTALL
+)
+
+
+def _cf_data_from_bytes(bytestring):
+    """
+    Given a bytestring, create a CFData object from it. This CFData object must
+    be CFReleased by the caller.
+    """
+    return CoreFoundation.CFDataCreate(
+        CoreFoundation.kCFAllocatorDefault, bytestring, len(bytestring)
+    )
+
+
+def _cf_dictionary_from_tuples(tuples):
+    """
+    Given a list of Python tuples, create an associated CFDictionary.
+    """
+    dictionary_size = len(tuples)
+
+    # We need to get the dictionary keys and values out in the same order.
+    keys = (t[0] for t in tuples)
```

Comment:
So, the answer is moderately complex.

A `CFDictionary` *in general* can hold pointers to truly arbitrary data. However, it needs a collection of functions it can use to help manage memory for them based on when they're added and removed. *This* `CFDictionary` is created with `kCFTypeDictionaryKeyCallBacks` and `kCFTypeDictionaryValueCallBacks`, which are the set that know how to manage memory for `CFTypeRef` objects (basically they do `CFRetain`/`CFRelease` as appropriate). That means that the keys and values need to be `CFTypeRef` objects: basically, any of the pointers that are handed to us by CoreFoundation or SecureTransport.

This is unfortunately really hard to enforce, because ctypes has a very confused understanding of the world. For that reason, we kinda just need to trust that the caller is doing the right thing here.